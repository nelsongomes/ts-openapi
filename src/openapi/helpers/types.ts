import Joi from "joi";

type Common = {
  description?: string;
  required?: boolean;
  nullable?: boolean;
};

type ParameterCommon = { isParameter?: boolean };
type ObjectCommon = { modelName?: string };

type StringParameters = {
  default?: string;
  example?: string;
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;

type StringEnumParameters = {
  values: string[];
  default?: string;
  example?: string;
} & Common &
  ParameterCommon;

type EmailParameters = StringParameters;
type UriParameters = StringParameters;
type HostnameParameters = StringParameters;
type PasswordParameters = {
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;
type UuidParameters = {
  example?: string;
} & Common &
  ParameterCommon;
type IpParameters = {
  example?: string;
} & Common &
  ParameterCommon;
type BinaryParameters = StringParameters;
type ByteParameters = StringParameters;

type DateParameters = {
  default?: string;
  example?: string;
} & Common &
  ParameterCommon;

type IntegerParameters = {
  default?: number;
  example?: number;
  minValue?: number;
  maxValue?: number;
} & Common &
  ParameterCommon;

type IntegerEnumParameters = {
  values: any[];
  default?: number;
  example?: number;
} & Common &
  ParameterCommon;

type NumberEnumParameters = IntegerEnumParameters & ParameterCommon;

type BooleanParameters = {
  default?: boolean;
  example?: boolean;
} & Common &
  ParameterCommon;

type ObjectParameters = {
  properties: Joi.SchemaMap<any>;
  default?: object;
  example?: object;
} & Common &
  ObjectCommon;

type ArrayParameters = {
  arrayType: Joi.SchemaLikeWithoutArray;
  default?: object;
  example?: object;
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;

export const Types = {
  String: (parameters?: StringParameters) => {
    let joiType = Joi.string();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        maxLength,
        minLength,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (minLength) {
        joiType = joiType.min(minLength);
      }

      if (maxLength) {
        joiType = joiType.max(maxLength);
      }

      if (parameters.default) {
        joiType = joiType.default(parameters.default);
      }

      if (parameters.example) {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  StringEnum: (parameters: StringEnumParameters) => {
    return Types.String(parameters).valid(...parameters.values);
  },

  Email: (parameters?: EmailParameters) => {
    return Types.String(parameters).email();
  },

  Password: (parameters?: PasswordParameters) => {
    return Types.String(parameters).meta({ format: "password" });
  },

  Uuid: (parameters?: UuidParameters) => {
    return Types.String(parameters)
      .uuid({ version: "uuidv4" })
      .meta({ format: "uuid" })
      .min(36)
      .max(36);
  },

  Uri: (parameters?: UriParameters) => {
    return Types.String(parameters)
      .uri()
      .meta({ format: "uri" });
  },

  Hostname: (parameters?: HostnameParameters) => {
    return Types.String(parameters)
      .hostname()
      .meta({ format: "hostname" });
  },

  Ipv4: (parameters?: IpParameters) => {
    return Types.String(parameters)
      .ip({ version: ["ipv4"] })
      .meta({ format: "ipv4" })
      .min(7)
      .max(15);
  },

  Ipv6: (parameters?: IpParameters) => {
    return Types.String(parameters)
      .ip({ version: ["ipv6"] })
      .meta({ format: "ipv6" })
      .min(3)
      .max(45);
  },

  Binary: (parameters?: BinaryParameters) => {
    let joiType = Joi.binary();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        maxLength,
        minLength,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (minLength) {
        joiType = joiType.min(minLength);
      }

      if (maxLength) {
        joiType = joiType.max(maxLength);
      }

      if (parameters.default) {
        joiType = joiType.default(parameters.default);
      }

      if (parameters.example) {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Byte: (parameters?: ByteParameters) => {
    return Types.Binary(parameters).encoding("base64");
  },

  DateTime: (parameters?: DateParameters) => {
    let joiType = Joi.string()
      .isoDate()
      .max(24);

    if (parameters) {
      const { description, required, nullable } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Date: (parameters?: DateParameters) => {
    return Types.DateTime(parameters)
      .meta({ format: "date" })
      .min(10)
      .max(10);
  },

  Number: (parameters?: IntegerParameters) => {
    let joiType = Joi.number();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        minValue,
        maxValue,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean" && required) {
        joiType = joiType.required();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (typeof minValue === "number") {
        joiType = joiType.min(minValue);
      }

      if (typeof maxValue === "number") {
        joiType = joiType.max(maxValue);
      }

      if (typeof parameters.default === "number") {
        joiType = joiType.default(parameters.default);
      }

      if (typeof parameters.example === "number") {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  NumberEnum: (parameters: NumberEnumParameters) => {
    return Types.Number(parameters).valid(...parameters.values);
  },

  Integer: (parameters?: IntegerParameters) => {
    return Types.Number(parameters).integer();
  },

  IntegerEnum: (parameters: IntegerEnumParameters) => {
    return Types.Integer(parameters).valid(...parameters.values);
  },

  Boolean: (parameters?: BooleanParameters) => {
    let joiType = Joi.boolean().strict();

    if (parameters) {
      const { description, required, nullable, example } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (typeof parameters.default === "boolean") {
        joiType = joiType.default(parameters.default);
      }

      if (typeof parameters.example === "boolean") {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Object: (parameters: ObjectParameters) => {
    const { properties, description, required, nullable, example } = parameters;
    let joiType = Joi.object().keys(properties);

    if (description) {
      joiType = joiType.description(description);
    }

    if (typeof required === "boolean") {
      joiType = required ? joiType.required() : joiType.optional();
    }

    if (nullable) {
      joiType = joiType.allow(null);
    }

    if (parameters.default) {
      joiType = joiType.default(parameters.default);
    }

    if (parameters.example) {
      joiType = joiType.example(example);
    }

    if (parameters.modelName) {
      joiType = joiType.meta({ modelName: parameters.modelName });
    }

    return joiType;
  },

  Array: (parameters: ArrayParameters) => {
    const {
      arrayType,
      description,
      required,
      nullable,
      example,
      minLength,
      maxLength
    } = parameters;
    let joiType = Joi.array().items(arrayType);

    if (description) {
      joiType = joiType.description(description);
    }

    if (typeof required === "boolean") {
      joiType = required ? joiType.required() : joiType.optional();
    }

    if (nullable) {
      joiType = joiType.allow(null);
    }

    if (parameters.default) {
      joiType = joiType.default(parameters.default);
    }

    if (parameters.example) {
      joiType = joiType.example(example);
    }

    if (minLength) {
      joiType = joiType.min(minLength);
    }

    if (maxLength) {
      joiType = joiType.max(maxLength);
    }

    if (parameters.isParameter) {
      joiType = joiType.meta({ parameter: true });
    }

    return joiType;
  }
};
