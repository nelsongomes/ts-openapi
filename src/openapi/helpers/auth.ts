import {
  SecurityApiKeyScheme,
  SecurityApiKeySchemeMethod,
  SecurityBasicScheme,
  SecurityBearerScheme,
  SecurityOauth2Scheme,
  SecurityOauth2Scopes,
} from "../openapi.types";

export function basicAuth(): SecurityBasicScheme {
  return {
    type: "http",
    scheme: "basic",
  };
}

export function apiKeyAuth(
  name: string,
  method: SecurityApiKeySchemeMethod
): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: method,
    name,
  };
}

export function cookieAuth(cookieName: string): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: "cookie",
    name: cookieName,
  };
}

export function bearerAuth(
  bearerFormat?: "JWT" | string
): SecurityBearerScheme {
  return {
    type: "http",
    scheme: "bearer",
    ...(bearerFormat && { bearerFormat }),
  };
}

export function oauth2ImplicitAuth(
  description: string,
  authorizationUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      implicit: {
        authorizationUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2AuthorizationCodeAuth(
  description: string,
  authorizationUrl: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      authorizationCode: {
        authorizationUrl,
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2PasswordAuth(
  description: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      password: {
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2ClientCredentialsAuth(
  description: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      clientCredentials: {
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}
