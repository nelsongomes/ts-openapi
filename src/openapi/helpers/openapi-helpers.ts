import {
  Body,
  SecurityApiKeyScheme,
  SecurityApiKeySchemeMethod,
  SecurityBasicScheme,
  SecurityBearerScheme,
} from "../openapi.types";

export function textPlain(description: string): Body {
  return {
    description,
    content: { "text-plain": {} },
  };
}

export function basicAuth(): SecurityBasicScheme {
  return {
    type: "http",
    scheme: "basic",
  };
}

export function apiKeyAuth(
  name: string,
  method: SecurityApiKeySchemeMethod
): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: method,
    name,
  };
}

export function cookieAuth(cookieName: string): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: "cookie",
    name: cookieName,
  };
}

export function bearerAuth(
  bearerFormat?: "JWT" | string
): SecurityBearerScheme {
  return {
    type: "http",
    scheme: "bearer",
    ...(bearerFormat && { bearerFormat }),
  };
}
