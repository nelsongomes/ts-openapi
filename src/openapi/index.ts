import Joi, { Schema } from "joi";

export { Joi, Schema };

export { OpenApi } from "./openapi";
export { WebRequestSchema, ParameterIn, Parameters } from "./openapi.types";

export { textPlain } from "./openapi-helpers";
