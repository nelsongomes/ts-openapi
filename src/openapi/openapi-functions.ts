import {
  IntegerFormats,
  NumberFormats,
  SchemaTypeBoolean,
  SchemaTypeInteger,
  SchemaTypeNumber,
  SchemaTypeString,
  StringFormats
} from "./openapi.types";

export function limitations(parameter: any): string {
  const limitationArray = [];

  if (parameter.type === "string" && parameter.enum) {
    // min & max should be ignored when is an enum
    delete parameter.minLength;
    delete parameter.maxLength;
  }

  // string
  if (parameter.minLength) {
    limitationArray.push(`min:${parameter.minLength} chars`);
  }

  if (parameter.maxLength) {
    limitationArray.push(`max:${parameter.maxLength} chars`);
  }

  // integer / float
  if (parameter.type === "integer" && parameter.enum) {
    // min & max should be ignored when is an enum
    delete parameter.minimum;
    delete parameter.maximum;
  }

  if (typeof parameter.minimum === "number") {
    limitationArray.push(`min:${parameter.minimum}`);
  }

  if (typeof parameter.maximum === "number") {
    limitationArray.push(`max:${parameter.maximum}`);
  }

  // array
  if (parameter.minItems) {
    limitationArray.push(`minItems:${parameter.minItems}`);
  }

  if (parameter.maxItems) {
    limitationArray.push(`maxItems:${parameter.maxItems}`);
  }

  if (parameter.format) {
    switch (parameter.format) {
      case StringFormats.DateTime:
        if (parameter.meta && parameter.meta.format === "date") {
          limitationArray.push(`date:yyyy-mm-dd`);
        } else {
          limitationArray.push(`ISO8601 date-time format`);
        }
        break;
      case StringFormats.Byte:
        limitationArray.push("base64 encoded string");
        break;
      case StringFormats.Binary:
        limitationArray.push("binary string");
        break;
    }
  }

  return limitationArray.length > 0 ? ` (${limitationArray.join(", ")})` : "";
}

export function openapiEscapeChars(key: string) {
  return key.replace(/~/g, "~0").replace(/\//g, "~1");
}

export function stringSchema(parameter: any): SchemaTypeString {
  const supportedFormats = Object.values(StringFormats);
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeString = {
    description,
    ...(parameter.default && { default: parameter.default }),
    ...(parameter.format &&
      supportedFormats.includes(parameter.format) && {
        format: parameter.format
      }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(parameter.minLength && { minLength: parameter.minLength }),
    ...(parameter.maxLength && { maxLength: parameter.maxLength }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "string"
  };

  if (parameter.meta) {
    if (parameter.meta.format) {
      parameter.format = parameter.meta.format;
    }
  }

  if (parameter.format && supportedFormats.includes(parameter.format)) {
    output.format = parameter.format;

    switch (parameter.format) {
      case StringFormats.Date:
        output.maxLength = 25; // 2021-11-05T00:00:00.000Z or 2020-10-14T21:44:03+00:00
        output.minLength = 10; // 2020-10-14
        break;
      case StringFormats.DateTime:
        output.minLength = 16; // 20201014T214403Z
        output.maxLength = 25; // 2021-11-05T00:00:00.000Z or 2020-10-14T21:44:03+00:00
        break;
      case StringFormats.Password:
        delete output.default; // no defaults for passwords
        break;
    }
  }

  // default values must be part of enum
  if (
    parameter.enum instanceof Array &&
    !parameter.enum.includes(output.default)
  ) {
    delete output.default;
  }

  return output;
}

export function numberSchema(parameter: any): SchemaTypeNumber {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeNumber = {
    description,
    ...(parameter.default && { default: parameter.default }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(typeof parameter.minimum === "number" && {
      minimum: parameter.minimum
    }),
    ...(typeof parameter.maximum === "number" && {
      maximum: parameter.maximum
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    format: NumberFormats.Double,
    type: "number"
  };

  return output;
}

export function booleanSchema(parameter: any): SchemaTypeBoolean {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeBoolean = {
    description,
    ...(typeof parameter.default === "boolean" && {
      default: parameter.default
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "boolean"
  };

  return output;
}

export function integerSchema(parameter: any): SchemaTypeInteger {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeInteger = {
    description,
    ...(typeof parameter.default === "number" && {
      default: parameter.default
    }),
    ...(typeof parameter.example === "number" && {
      example: parameter.example
    }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(typeof parameter.minimum === "number" && {
      minimum: parameter.minimum
    }),
    ...(typeof parameter.maximum === "number" && {
      maximum: parameter.maximum
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    format: IntegerFormats.Int64,
    type: "integer"
  };

  // default values must be part of enum
  if (
    parameter.enum instanceof Array &&
    !parameter.enum.includes(output.default)
  ) {
    delete output.default;
  }

  return output;
}
