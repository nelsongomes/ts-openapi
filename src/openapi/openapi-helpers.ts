export function textPlain(description: string) {
  return {
    content: {
      "text/plain": {
        schema: {
          type: "string",
        },
      },
    },
    description,
  };
}
