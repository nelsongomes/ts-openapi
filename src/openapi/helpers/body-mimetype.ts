import { Body } from "../openapi.types";

export function textPlain(description: string): Body {
  return {
    description,
    content: { "text-plain": {} },
  };
}
