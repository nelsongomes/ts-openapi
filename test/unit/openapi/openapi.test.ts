import { OpenApi } from "../../../src/openapi/openapi";
import { textPlain } from "../../../src/openapi/helpers/body-mimetype";
import { Types } from "../../../src/openapi/helpers/types";

describe("src/openapi/openapi", () => {
  describe("Constructor", () => {
    test("simple", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      expect(openApi instanceof OpenApi).toBe(true);
    });

    test("add body", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      const body = Types.Object({
        properties: {
          username: Types.String()
        },
        description: "Sample body"
      });

      openApi.setServers([{ url: "https://server.com" }]);
      openApi.addPath(
        "/something",
        {
          post: {
            operationId: "id",
            description: "desc",
            summary: "summary",
            tags: ["example"],
            responses: {
              [200]: { description: "Success", content: { "plain/text": {} } }
            },
            validationSchema: { body }
          }
        },
        true
      );

      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("simple, but no servers were added", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.generateJson();
        fail("Should have thrown exception");
      } catch (e) {
        expect(e.message).toBe("No servers were added to OpenApi definition");
      }
    });

    test("simple, but no paths were added", done => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);
        openApi.generateJson();

        done.fail("Should have thrown exception");
      } catch (e) {
        expect(e.message).toBe("No paths were added to OpenApi definition.");

        done();
      }
    });

    test("simple, added servers and paths", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);
        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "healthcheck",
              validationSchema: {},
              responses: {
                200: textPlain("Successful operation.")
              },
              summary: "Server Healthcheck",
              tags: ["Internals"]
            }
          },
          true
        );

        expect(openApi.generateJson()).toMatchSnapshot();
      } catch (e) {
        fail("No exception expected");
      }
    });

    test("simple, added servers and paths, get cannot have body", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      const body = Types.Object({
        properties: {
          username: Types.String()
        },
        description: "Sample body"
      });

      try {
        openApi.setServers([{ url: "https://server.com" }]);
        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "healthcheck",
              validationSchema: { body },
              responses: {
                200: textPlain("Successful operation.")
              },
              summary: "Server Healthcheck",
              tags: ["Internals"]
            }
          },
          true
        );

        openApi.generateJson();
      } catch (e) {
        expect(e.message).toBe("GET operations cannot have a requestBody.");
      }
    });

    test("addPath, no operationId", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);
        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              validationSchema: {},
              responses: {
                200: textPlain("Successful operation.")
              },
              summary: "Server Healthcheck",
              tags: ["Internals"]
            } as any
          },
          true
        );

        fail("Expected to throw exception");
      } catch (e) {
        expect(e.message).toBe("No operationId supplied.");
      }
    });

    test("addPath, no response", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);
        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "",
              validationSchema: {},
              responses: {},
              summary: "Server Healthcheck",
              tags: ["Internals"]
            }
          },
          true
        );

        fail("Expected to throw exception");
      } catch (e) {
        expect(e.message).toBe("Should define at least one response.");
      }
    });

    test("addPath, duplicate operationId", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);
        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "repeated",
              validationSchema: {},
              responses: {
                200: textPlain("Successful operation.")
              },
              summary: "Server Healthcheck",
              tags: ["Internals"]
            }
          },
          true
        );

        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "repeated",
              validationSchema: {},
              responses: {
                200: textPlain("Successful operation.")
              },
              summary: "Server Healthcheck",
              tags: ["Internals"]
            }
          },
          true
        );

        fail("Expected to throw exception");
      } catch (e) {
        expect(e.message).toBe("Operations must have unique operationIds.");
      }
    });

    test("addPath, public vs private method (private method does not get included)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      openApi.setServers([{ url: "https://server.com" }]);
      openApi.addPath(
        "/health",
        {
          get: {
            description: "Service healthcheck endpoint",
            operationId: "repeated",
            validationSchema: {},
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Healthcheck",
            tags: ["Internals"]
          }
        },
        true
      );

      openApi.addPath(
        "/health",
        {
          get: {
            description: "Service healthcheck endpoint",
            operationId: "repeated",
            validationSchema: {},
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Healthcheck",
            tags: ["Internals"]
          }
        },
        false
      );

      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("setLicense", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      openApi.setServers([{ url: "https://server.com" }]);
      openApi.addPath(
        "/health",
        {
          get: {
            description: "Service healthcheck endpoint",
            operationId: "repeated",
            validationSchema: {},
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Healthcheck",
            tags: ["Internals"]
          }
        },
        true
      );

      openApi.setLicense(
        "Apache 33.0",
        "http://www.apache.org/licenses/LICENSE-33.0.html",
        "http://dummy.io/terms/"
      );

      expect(openApi.generateJson()).toMatchSnapshot();
    });
  });
});
