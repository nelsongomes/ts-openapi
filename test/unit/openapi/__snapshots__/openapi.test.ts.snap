// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/openapi/openapi Constructor add body 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/something": Object {
      "patch": Object {
        "description": "desc",
        "operationId": "patchid",
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "description": "Sample body",
                "properties": Object {
                  "username": Object {
                    "description": "Parameter without description.",
                    "type": "string",
                  },
                },
                "type": "object",
              },
            },
          },
          "description": "Sample body",
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "plain/text": Object {},
            },
            "description": "Success",
          },
        },
        "summary": "summary",
        "tags": Array [
          "example",
        ],
      },
      "post": Object {
        "description": "desc",
        "operationId": "postid",
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "description": "Sample body",
                "properties": Object {
                  "username": Object {
                    "description": "Parameter without description.",
                    "type": "string",
                  },
                },
                "type": "object",
              },
            },
          },
          "description": "Sample body",
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "plain/text": Object {},
            },
            "description": "Success",
          },
        },
        "summary": "summary",
        "tags": Array [
          "example",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Constructor add body, referenced 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "Username": Object {
        "description": "Sample body",
        "properties": Object {
          "username": Object {
            "description": "Parameter without description.",
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/something": Object {
      "patch": Object {
        "description": "desc",
        "operationId": "patchid",
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/Username",
              },
            },
          },
          "description": "Sample body",
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "plain/text": Object {},
            },
            "description": "Success",
          },
        },
        "summary": "summary",
        "tags": Array [
          "example",
        ],
      },
      "post": Object {
        "description": "desc",
        "operationId": "postid",
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/Username",
              },
            },
          },
          "description": "Sample body",
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "plain/text": Object {},
            },
            "description": "Success",
          },
        },
        "summary": "summary",
        "tags": Array [
          "example",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Constructor addPath, public vs private method (private method does not get included) 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Constructor setLicense 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 33.0",
      "url": "http://www.apache.org/licenses/LICENSE-33.0.html",
    },
    "termsOfService": "http://dummy.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Constructor simple, added servers and paths 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "healthcheck",
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;
